File: shopping-cart-service-old/src/cart/domain/cart-item.ts
import { Money } from './value-objects/money';
import { Quantity } from './value-objects/quantity';
import { randomUUID } from 'crypto';

export class CartItem {
  private constructor(
    public id: string,
    public productId: string,
    public productName: string,
    public price: Money,
    public quantity: Quantity,
  ) {}

  static create(
    productId: string,
    productName: string,
    price: Money,
    quantity: Quantity,
  ): CartItem {
    return new CartItem(randomUUID(), productId, productName, price, quantity);
  }

  updateQuantity(quantity: Quantity): void {
    this.quantity = quantity;
  }

  updatePrice(price: Money): void {
    this.price = price;
  }

  subtotal(): Money {
    return this.price.multiply(this.quantity.value);
  }
}

File: shopping-cart-service/src/main/java/com/carlosoliveira/ecommerce/shoppingcart/domain/CartItem.java
package com.carlosoliveira.ecommerce.shoppingcart.domain;

import com.carlosoliveira.ecommerce.common.valueObjects.Money;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.UUID;

@Entity
@Table(name = "cart_items")
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class CartItem {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column
    private UUID productId;

    @Column
    private String productName;

    @Column
    private int quantity;

    @Embedded
    private Money price;

    @ManyToOne
    @JoinColumn(name = "cart_id")
    private Cart cart;

    public CartItem(
            UUID productId,
            String productName,
            Money price,
            int quantity
    ) {
        this.productId = productId;
        this.productName = productName;
        this.price = price;
        this.quantity = quantity;
    }

    public void updateQuantity(int quantity) {
        if (quantity <= 0) {
            throw new IllegalArgumentException("Quantity must be greater than zero.");
        }
        this.quantity = quantity;
    }

    public void updatePrice(Money price) {
        this.price = price;
    }

    public Money subtotal() {
        return this.price.multiply(this.quantity);
    }
}
